doc/book/box/indexes.rst

```diff
@@ -24,46 +20,453 @@ The first index defined on a space is called the **primary key index**,
 
 The effect is that, for all tuples in space 'tester', field #2 must exist and
 must contain a string.
-The index type is 'tree', so values in field #2 must not be unique, because keys
+The index type is TREE, so values in field #2 must not be unique, because keys
 in TREE indexes may be non-unique.
 
-.. NOTE::
+Space definitions and index definitions are stored permanently in Tarantool's
+system spaces :ref:`_space <box_space-space>` and :ref:`_index <box_space-index>`
+(for details, see reference on :ref:`box.space <box_space>` submodule).
+
+..  admonition:: Tip
+    :class: fact
+
+    The full information about creating an index is in section
+    :doc:`/reference/reference_lua/box_space/create_index`.
+
+--------------------------------------------------------------------------------
+Index types
+--------------------------------------------------------------------------------
+
+An index always has a **type**. Different types are intended for different
+usage scenarios.
+
+We give an overview of index features in the following table:
+
+..  container:: table
+
+    ..  list-table::
+        :header-rows: 1
+
+        *   - Feature
+            - TREE
+            - HASH
+            - RTREE
+            - BITSET
+
+        *   - unique
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - non-unique
+            - \+
+            - \-
+            - \+
+            - \+
+
+        *   - :ref:`is_nullable <box_space-is_nullable>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - multi-part
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - :ref:`multikey <box_space-path_multikey>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - :ref:`partial-key search <partial_key_search>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - can be primary key
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - exclude_null (version 2.8+)
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - :doc:`iterator types </reference/reference_lua/box_index/pairs>`
+            - ALL, EQ, REQ, GT, GE, LT, LE
+            - ALL, EQ, GT
+            - ALL, EQ, GT, GE, LT, LE, OVERLAPS, NEIGHBOR
+            - ALL, EQ, BITS_ALL_SET, BITS_ANY_SET, BITS_ALL_NOT_SET
+
+********************************************************************************
+TREE indexes
+********************************************************************************
+
+The default index type is 'TREE'.
+TREE indexes are provided by memtx and vinyl engines, can index unique and
+non-unique values, support partial key searches, comparisons and ordered results.
+
+This is a universal type of indexes, for most cases it will be the best choice.
+
+Additionally, memtx engine supports HASH, RTREE and BITSET indexes.
```

Comment:
If vinyl doesn't support other types, write it in the sections about other types.
Or maybe it can fit in the comparison table above.